Task ownership and access-control

mrtamm opened the pull request:
This development provides configurable access-control for tasks, and introduces optional admin-user role for managing everyone's tasks. To enable ownership, each database has to record the owner (username) when task is created, and check ownership (if required by the configured access-mode) each time data is requested or updated. Meanwhile, the system has its own system-user, which has elevated access-permission for performing its operations (computation event-handling).

Through this feature we would like to restrict visibility and access to tasks in a multi-user service. On the other hand, the configuration retains the current access-mode option where authenticated users can see and access all tasks.

There are also documentation updates regarding access-control configuration and the new fields of database configuration.

I tested the new functionality with all possible databases (badger, bolt, datastore, dynamodb, elastic, and mongo) to be sure that the queries perform correctly.

Although this access-control proposal provides a working solution, any feedback and suggestion is still welcome to handle non-covered user-needs.

## Configuration changes

1. Basic-authentication users now have optional `Admin: true` property
2. OIDC-authentication now has optional `Admins` property (list of usernames)
3. `Server.TaskAccess` with options `All`, `Owner`, `OwnerOrAdmin`. For compatibility, the default is `All`, which is the current behaviour.
4. `Elastic` configuration now has more fields but all the new ones are optional:
   * `Username`
   * `Password`
   * `CloudID`
   * `APIKey`
   * `ServiceToken`

## Badger DB changes

* `owner` is stored under separate key `owners<task_id>`
* the owner is read every time when tasks are accessed

## Bolt DB changes

* `owner` is stored in a separate bucket: `tasks-owner`
* the owner is read every time when tasks are accessed and the access-mode implies checking for the owner

## Datastore changes

* `owner` is stored in the `Task` entity: under the `Owner` field.
* if access-mode implies it, the `Owner` filter is applied in the list-tasks query.
* when retrieving a single Task, the owner-check compares the field-value.
* I had to make more code changes to reading and writing the `Task` entity so that the original `owner` would not be overwritten.
* I added the Datastore-specific **index.yaml** file which describes required indexes to support the list-tasks query.

## DynamoDB changes

* `owner` is stored in the `task` table: in the `owner` column.
* if access-mode implies it, the `owner` filter is applied in the list-tasks query.
* when retrieving a single Task, the owner-check compares the column-value.

## Elastic changes

* I had to upgrade the client-library, as the current one supported v5 while the latest Elasticsearch is v8. Fortunately, the driver API did not change much.
* `owner` is stored in the `task` index: in the `owner` property.
* if access-mode implies it, the `owner` filter is applied in the list-tasks query.
* when retrieving a single Task, the owner-check compares the field-value.

## MongoDB changes

* I had to upgrade to v2 Mongo driver, which implied some API changes
* I updated requests with timeout contexts to apply time-restrictions
* `owner` is stored in the `task` collection: in the `owner` property.
* if access-mode implies it, the `owner` filter is applied in the list-tasks query.

kellrott commented:
Wow. Thank you. This will take some work to review but we'll get started on it as soon as possible.

lbeckman314 commented:
Thank you Martii! We will be testing this PR tomorrow and hopefully then getting it into the next official release cycle 🌀 

# Testing Steps ✍️ 

```sh
<REPLACE ME>
```

mrtamm commented:
Thank you! I did some additional testing myself, and fixed some issues (pushed to the branch).

Good news is that I created a Bash script for automating my tests, as I needed to perform these steps for each database. Overview of my configuration, docker-compose (for local Mongo and Elasticsearch), and the test-workflow-script are provided in this gist: https://gist.github.com/mrtamm/f97da6e7787083bbdd5897cf6b037944. I hope it helps.

The test-script focuses only on `TaskAccess: OwnerOrAdmin` as other modes are similar: `All` permits all, and `Owner` restricts to owners only (which is a subset of `OwnerOrAdmin` test-cases). As shown in the configuration, it uses 3 Basic-authentication users: two standard ones and one Admin. All requests are triggered from command-line `funnel task list|get|cancel`.

lbeckman314 commented:
Thanks Martti! Just ran an initial test pass and things are looking good — I will revisit the error I encountered above when running the test script first thing Monday (_likely user error on my part_) and report back! 🌊

mrtamm commented:
Thank you for the feedback!

After another round of testing, I did some minor updates again:

1. merged incoming changes from the `develop` branch + updated dependencies (go mod tidy)
2. fixed an error in the datastore data loading: without any logs `CustomMarshal.DetectView()` failed
3. Fixed `tes.GetPageSize()` to ignore non-positive values
4. Improved MongoDB `ListTasks` response to skip the `nextPageToken` from the output JSON when there are definitely no more items to load.
5. Though it's not related to current feature, I fixed the failing test in `tes/utils_test.go`. It was expecting the output from `tes.Base64Encode()` to match the expected string. However, [the protojson marshaller intentionally adds random (but safe) whitespace to JSON](https://github.com/golang/protobuf/issues/1082), so BASE64-encoded output varies with each run. As the test was more about checking the `Base64Decode()` method, I just removed the intermediary check of the BASE64 string. Now the tests are green again. 👍

In addition, I updated the [test-flow-script.sh](https://gist.github.com/mrtamm/f97da6e7787083bbdd5897cf6b037944#file-test-flow-script-sh) script:
1. prints `nextPageToken` values when they are present in the output of `task list`
2. only checks for `systemLogs` when the task is COMPLETE
3. reduced the pause between create-task and cancel-task (it failed when the task managed to complete quickly)

Please try the updated test-flow script, since I noticed that you had an error related to the `systemLogs` check.

I did not notice any case where `task list` would return an empty (`""`) `nextPageToken` (#1138). If it's still an issue, I might need more precise config/setup description for reproducing it.

lbeckman314 commented:
Thank you Martii! We will be testing this PR tomorrow and hopefully then getting it into the next official release cycle 🌀 

# Testing Steps ✍️ 

> Adapted from [testing steps](https://gist.github.com/mrtamm/f97da6e7787083bbdd5897cf6b037944) by @mrtamm 

## 1. Installing PR #1137 ✅ 

```sh
➜ gh pr checkout 1137
branch 'develop-task-ownership' set up to track 'KrKOo/develop-task-ownership'.
Switched to a new branch 'develop-task-ownership'

➜ make install

➜  funnel
$HOME/go/bin/funnel

➜ funnel version
git commit:  6db8a5bd
git branch: develop-task-ownership
git upstream: https://github.com/mrtamm/funnel.git
build date: Fri Jan 31 18:44:36 PST 2025
version: unknown
```

## 2. Start Funnel Server (with Minimal Local Test Config) ✅ 

```sh
➜ cat <<EOF > funnel-test-config.yaml
Compute: Docker

Server:
  HostName: localhost
  HTTPPort: 8000
  RPCPort: 9090
  BasicAuth:
    - User: admin
      Password: admin
      Admin: true
    - User: user1
      Password: user-one-pass
    - User: user2
      Password: user-two-pass
    - User: user-three
      Password: user-three-pass
  TaskAccess: OwnerOrAdmin
EOF

➜ funnel server run --config funnel-test-config.yaml
```

## 3. Start Docker Compose ✅ 

```sh
➜ docker compose --file funnel-docker-compose.yaml up --detatch
```

## 4. Run Tasks ✅ 

```sh
# Unauthenticated
➜ funnel task list
Error: [STATUS CODE - 401]      {
  "error": "Basic-authentication failed"
}

# Authenticated (user 1)
➜ export FUNNEL_SERVER_USER=user1
➜ export FUNNEL_SERVER_PASSWORD=user-one-pass
➜ funnel task list
{}

➜ funnel examples hello-world > hello-world.json

➜ funnel task create hello-world.json
cuep0v1urbu7e04d2k8g

➜ funnel task list
{
  "tasks":  [
    {
      "creationTime":  "2025-01-31T19:06:49.20143-08:00",
      "description":  "Demonstrates the most basic echo task.",
      "executors":  [
        {
          "command":  [
            "echo",
            "hello world"
          ],
          "image":  "alpine"
        }
      ],
      "id":  "cuep0v1urbu7e04d2k8g",
      "logs":  [
        {}
      ],
      "name":  "Hello world",
      "state":  "QUEUED"
    }
]

# Authenticated (User 2)
➜ export FUNNEL_SERVER_USER=user2
➜ export FUNNEL_SERVER_PASSWORD=user-two-pass
➜  funnel task list
{}

# Mixing credentials (expected 401 error)
➜ export FUNNEL_SERVER_USER=user1
➜ export FUNNEL_SERVER_PASSWORD=user-two-pass
➜  funnel git:(develop-task-ownership) ✗ funnel task list
Error: [STATUS CODE - 401]      {
  "error": "Basic-authentication failed"
}
```

## 5. Dashboard Authentication ✅ 

> Initial dashboard auth prompt (with successful login with admin credentials specified in Funnel's config file)

![Dashboard Auth](https://github.com/user-attachments/assets/29600556-4fe6-40a9-8781-04c570c24e26)

> Logging in as `user1` from above

![Screenshot 2025-01-31 at 7 13 41 PM](https://github.com/user-attachments/assets/5da0bd87-559b-42e9-b376-ed321d8663ab)

## 6. Run Test Script ⚠️ 

```sh
➜ chmod +x ./test-flow-script.sh

➜ cp $(which funnel) .

➜ ./test-flow-script.sh
[user1]> ./funnel task list   # Expecting 0 task(s)
[user1]> ./funnel task create test-task.json  # -> task ID: cuep541urbufv4k3vm80
[user1]> ./funnel task cancel cuep541urbufv4k3vm80   # Expecting success
[user1]> ./funnel task list -v minimal  # Expecting 1 task(s)
[user1]> ./funnel task list -v basic  # Expecting 1 task(s)
[user1]> ./funnel task list -v full  # Expecting 1 task(s)
FAIL: did not find systemLogs in a task {
  "tasks":  [
    {
      "creationTime":  "2025-01-31T19:18:40.286052-08:00",
      "description":  "For testing.",
      "executors":  [
        {
          "command":  [
            "echo",
            "Hello from the container"
          ],
          "image":  "alpine"
        }
      ],
      "id":  "cuep541urbufv4k3vm80",
      "logs":  [
        {}
      ],
      "name":  "Basic TEST Task",
      "state":  "CANCELED",
      "tags":  {
        "test":  "true"
      }
    }
  ]
}
```

# Next Steps: ⚡️
> OHSU's team will resolve these ahead of merge! 🌊 
- [ ] Resolve the Failed Test Script run above (`failed to get task. ID unknown` error from Server) when submitting tasks (_likely user error on tester's part!_)
- [ ] Resolve `go.mod` and `go.sum` conflicts
- [ ] Fix K8s and Slurm tests (underlying Funnel issue, not part of this PR)

lbeckman314 commented:
Ah copy that! And looks good — running the test script again with the issue I was encountering before now resolved* and updating the test steps above ✍️ 

## Issue + Resolution ✅ 

1. The `failed to get task. ID unknown` issue was due to me not having the correct `RPCClient` credentials in the config, 

> Ref: https://ohsu-comp-bio.github.io/funnel/docs/security/basic/

```yaml
RPCClient:
  User: funnel
  Password: abc123
```

2. Which then resulted a `Basic-authentication failed` RPC error on the Server side (due to not having these credentials set in the config):

https://github.com/ohsu-comp-bio/funnel/blob/f704f446d4de4801119cb65133c9543af83875ff/util/rpc/rpc.go#L16-L23

3. This was then getting "lost" in the following Worker error message:

https://github.com/ohsu-comp-bio/funnel/blob/f704f446d4de4801119cb65133c9543af83875ff/util/rpc/rpc.go#L16-L23

lbeckman314 reviewed (approved):
All tests passing perfectly and code looks great — 100% approved! 🚀

All local testing steps recorded in this PR comment:
> https://github.com/ohsu-comp-bio/funnel/pull/1137#issuecomment-2623311866

## Next Steps to Merge and Release 🌀 
> To be completed by OHSU Team by EOD Tuesday March 4th ⏰ 

- [ ] @lbeckman314 update the four failing tests below that are not recognizing new fields in the config?
- [ ] Get OK with @kellrott to merge to `develop` and cut new official release

mrtamm commented:
Thanks again! I'm glad that it fits into the next release :) 

As I went through the test-scenarios provided in the Makefile, I discovered a bug where the metadata was not stored to Datastore, so I had to make one more commit to fix that.

---

Other changes I had to make to fix the Makefile-tests were in the source code of the tests themselves, and not in that of the developed features. I have not committed these changes as I would prefer resolving the tests in another PR but if you prefer otherwise I can also push them here.❓

A typical problem with the tests (on some databases) was data consistency: getting a single task was more accurate than searching and listing tasks. Adding some small pauses helped.

mrtamm commented:
I was debugging why the **slurmTest** failed and found out that it was using a Funnel build from the master branch:

```
# ./funnel version                 
git commit:  a26c1c5
git branch: master
git upstream: https://github.com/ohsu-comp-bio/funnel
build date: Wed Jan  8 22:32:31 UTC 2025
version: unknown
```

It explains why it failed to parse the generated configuration file.

Most likely this issue is related to the way the **build** job does file-caching (the last line under `restore-keys` looks suspicious):

```yaml
    - name: Cache Funnel binary
      uses: actions/cache@v3
      with:
        path: ./funnel
        key: ${{ runner.os }}-funnel-bin-${{ hashFiles('**/go.sum') }}-${{ github.ref }}
        restore-keys: |
          ${{ runner.os }}-funnel-bin-${{ github.ref }}
          ${{ runner.os }}-funnel-bin-
```

Reference to the source:
https://github.com/mrtamm/funnel/blob/d4601c994093a73913211b5e199d61c56f1d2817/.github/workflows/build.yml#L19

If you have any hints, I could also include a fix for this.

My journey with the unit tests is coming to an end and I will commit them tomorrow (to a separate branch for better testing on GitHub).

mrtamm commented:
The changes to the pipelines and tests (also covering the `TaskAccess` parameter) are currently in this PR: https://github.com/mrtamm/funnel/pull/1. @lbeckman314 please have look, if the changes are fine. (I also wrote there why I could not resolve the Kubernetes tests.)

If there are no issues with the changes, I will merge them here tomorrow. After that, this PR will be complete.

lbeckman314 commented:
Looks good to me! We can safely ignore the K8s Test as that's an unrelated (and non-blocking) issue currently being addressed in PR #1074 

Excited to have this feature formally introduced — thanks again for the great code, tests, and documentation @mrtamm!
